File: backend/app/schemas/knowledge.py
"""Knowledge Base schemas for request/response validation."""

from typing import Optional

from pydantic import BaseModel


class CreateKBRequest(BaseModel):
    title: str
    description: Optional[str] = None


class UpdateKBRequest(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    shared_with_team: Optional[bool] = None
    organization_ids: Optional[list[str]] = None


class AddDocumentsRequest(BaseModel):
    document_uuids: list[str]


class ConvertDocumentsRequest(BaseModel):
    """Wrap one or more SmartDocuments in a new KB so they can be retrieved
    instead of inlined. Used by the chat / workflow "Convert to Knowledge
    Base" affordance shown when a doc is too large for the current model.
    """
    document_uuids: list[str]
    title: Optional[str] = None  # defaults to the first doc's title


class ShareKBRequest(BaseModel):
    comment: Optional[str] = None


class AddUrlsRequest(BaseModel):
    urls: list[str]
    crawl_enabled: bool = False
    max_crawl_pages: int = 5
    allowed_domains: str = ""  # comma-separated


class KBSourceResponse(BaseModel):
    uuid: str
    source_type: str
    document_uuid: Optional[str] = None
    document_title: Optional[str] = None  # filename, looked up at response time
    url: Optional[str] = None
    url_title: Optional[str] = None
    custom_name: Optional[str] = None  # user-provided label; UI prefers this over title/url
    status: str
    error_message: Optional[str] = None
    chunk_count: int = 0
    created_at: Optional[str] = None


class UpdateSourceRequest(BaseModel):
    """Patch a single KB source. Empty string clears the custom name."""
    custom_name: Optional[str] = None


class KBResponse(BaseModel):
    uuid: str
    title: str
    description: Optional[str] = None
    status: str
    shared_with_team: bool = False
    team_owned: bool = False
    verified: bool = False
    organization_ids: list[str] = []
    total_sources: int = 0
    sources_ready: int = 0
    sources_failed: int = 0
    total_chunks: int = 0
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    # Scope & ownership fields for the UI
    user_id: Optional[str] = None
    scope: Optional[str] = None  # "mine" | "team" | "verified" | "reference"
    is_reference: bool = False
    source_kb_uuid: Optional[str] = None  # set when is_reference=True
    reference_uuid: Optional[str] = None  # the reference's own uuid


class KBListResponse(BaseModel):
    items: list[KBResponse] = []
    total: int = 0


class AdoptKBRequest(BaseModel):
    note: Optional[str] = None
    team_id: Optional[str] = None  # adopt to a specific team (default: personal)


class KBReferenceResponse(BaseModel):
    uuid: str
    source_kb_uuid: str
    user_id: str
    team_id: Optional[str] = None
    note: Optional[str] = None
    pinned: bool = False
    created_at: Optional[str] = None


class KBDetailResponse(KBResponse):
    sources: list[KBSourceResponse] = []


class KBStatusResponse(BaseModel):
    uuid: str
    status: str
    total_sources: int = 0
    sources_ready: int = 0
    sources_failed: int = 0
    total_chunks: int = 0
    sources: list[dict] = []


# --- Export / Import ---

KB_EXPORT_FORMAT_VERSION = 1


class KBExportSource(BaseModel):
    source_type: str  # "document" | "url"
    document_uuid: Optional[str] = None
    document_title: Optional[str] = None  # snapshot of SmartDocument.title at export time
    url: Optional[str] = None
    url_title: Optional[str] = None
    custom_name: Optional[str] = None  # user's chosen label, carried across export/import
    content: Optional[str] = None  # cached raw text (for URLs) or document raw_text (for docs)
    crawl_enabled: bool = False
    max_crawl_pages: int = 5
    parent_source_uuid: Optional[str] = None
    crawled_urls: Optional[list[str]] = None


class KBExportPayload(BaseModel):
    format_version: int = KB_EXPORT_FORMAT_VERSION
    exported_at: Optional[str] = None
    title: str
    description: Optional[str] = None
    sources: list[KBExportSource] = []


class ImportKBRequest(BaseModel):
    payload: KBExportPayload
    title: Optional[str] = None  # override title on import


class ImportKBResponse(BaseModel):
    uuid: str
    title: str
    imported_sources: int

File: frontend/src/types/knowledge.ts
export type KBScope = 'mine' | 'team' | 'verified' | 'reference'

export interface KnowledgeBase {
  uuid: string
  title: string
  description: string
  status: 'empty' | 'building' | 'ready' | 'error'
  shared_with_team: boolean
  team_owned: boolean
  verified: boolean
  organization_ids: string[]
  tags: string[]
  team_id: string | null
  total_sources: number
  sources_ready: number
  sources_failed: number
  total_chunks: number
  created_at: string
  updated_at: string
  // Scope & ownership fields (from v2 list endpoint)
  user_id?: string
  scope?: KBScope
  is_reference?: boolean
  source_kb_uuid?: string
  reference_uuid?: string
  // Set by KB Autovalidate's apply path
  has_optimized_config?: boolean
  optimized_config_set_at?: string | null
  // AI-trust signals from the latest validation run. Scores are 0-1.
  last_validation_score?: number | null
  last_validation_baseline_score?: number | null
  last_validation_lift?: number | null
  last_validated_at?: string | null
}

export interface KnowledgeBaseSource {
  uuid: string
  source_type: 'document' | 'url'
  document_uuid?: string
  document_title?: string
  url?: string
  url_title?: string
  custom_name?: string | null
  status: 'pending' | 'processing' | 'ready' | 'error'
  error_message?: string
  chunk_count: number
  created_at: string
}

export interface KnowledgeBaseSourceDetail extends KnowledgeBaseSource {
  content?: string | null
  crawl_enabled: boolean
  max_crawl_pages: number
  parent_source_uuid?: string | null
  crawled_urls?: string[] | null
  child_sources: KnowledgeBaseSource[]
  processed_at?: string | null
}

export interface KnowledgeBaseDetail extends KnowledgeBase {
  sources: KnowledgeBaseSource[]
}

export interface KBListResponse {
  items: KnowledgeBase[]
  total: number
}

export interface KBReference {
  uuid: string
  source_kb_uuid: string
  user_id: string
  team_id?: string
  note?: string
  pinned: boolean
  created_at?: string
}
